src/esp/gfx/GenericDrawable.cpp

```diff
@@ -4,43 +4,106 @@
 
 namespace esp {
 namespace gfx {
 
-GenericDrawable::GenericDrawable(
-    scene::SceneNode& node,
-    Magnum::Shaders::Flat3D& shader,
-    Magnum::GL::Mesh& mesh,
-    DrawableGroup* group /* = nullptr */,
-    Magnum::GL::Texture2D* texture /* = nullptr */,
-    int objectId /* = ID_UNDEFINED */,
-    const Magnum::Color4& color /* = Magnum::Color4{1} */)
-    : Drawable{node, shader, mesh, group},
-      texture_(texture),
-      objectId_(objectId),
-      color_{color} {}
+GenericDrawable::GenericDrawable(scene::SceneNode& node,
+                                 Magnum::GL::Mesh& mesh,
+                                 ShaderManager& shaderManager,
+                                 const Magnum::ResourceKey& lightSetup,
+                                 const Magnum::ResourceKey& materialData,
+                                 DrawableGroup* group /* = nullptr */,
+                                 int objectId /* = ID_UNDEFINED */)
+    : Drawable{node, mesh, group},
+      shaderManager_{shaderManager},
+      lightSetup_{shaderManager.get<LightSetup>(lightSetup)},
+      materialData_{
+          shaderManager.get<MaterialData, PhongMaterialData>(materialData)},
+      objectId_(objectId) {
+  // update the shader early here to to avoid doing it during the render loop
+  updateShader();
+}
+
+void GenericDrawable::setLightSetup(const Magnum::ResourceKey& resourceKey) {
+  lightSetup_ = shaderManager_.get<LightSetup>(resourceKey);
+
+  // update the shader early here to to avoid doing it during the render loop
+  updateShader();
+}
 
 void GenericDrawable::draw(const Magnum::Matrix4& transformationMatrix,
                            Magnum::SceneGraph::Camera3D& camera) {
-  Magnum::Shaders::Flat3D& shader =
-      static_cast<Magnum::Shaders::Flat3D&>(shader_);
-  shader.setTransformationProjectionMatrix(camera.projectionMatrix() *
-                                           transformationMatrix);
+  updateShader();
+
+  (*shader_)
+      .setAmbientColor(materialData_->ambientColor)
+      .setDiffuseColor(materialData_->diffuseColor)
+      .setSpecularColor(materialData_->specularColor)
+      .setShininess(materialData_->shininess)
+      .setObjectId(node_.getId())
+      .setTransformationMatrix(transformationMatrix)
+      .setProjectionMatrix(camera.projectionMatrix())
+      .setNormalMatrix(transformationMatrix.rotationScaling());
+
+  if (materialData_->ambientTexture)
+    shader_->bindAmbientTexture(*(materialData_->ambientTexture));
+  if (materialData_->diffuseTexture)
+    shader_->bindDiffuseTexture(*(materialData_->diffuseTexture));
+  if (materialData_->specularTexture)
+    shader_->bindSpecularTexture(*(materialData_->specularTexture));
+
+  const Magnum::Matrix4 cameraMatrix = camera.cameraMatrix();
+  for (Magnum::UnsignedInt i = 0; i < lightSetup_->size(); ++i) {
+    shader_->setLightPosition(
+        i, getLightPositionRelativeToCamera(
+               (*lightSetup_)[i], transformationMatrix, cameraMatrix));
```

Comment:
To reduce the number of GL calls it might be better to first calculate all positions to a pre-sized array and then call `setLightPositions()` / `setLightColors()` for all lights at once.